spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java

```diff
@@ -183,8 +180,8 @@ else if (StringUtils.hasText(asyncExecutor)) {
 
 		gatewayProxyBuilder.addConstructorArgValue(serviceInterface);
 
-		AbstractBeanDefinition beanDefinition = gatewayProxyBuilder.getBeanDefinition();
-		beanDefinition.setAttribute(FactoryBean.OBJECT_TYPE_ATTRIBUTE, serviceInterface);
+		RootBeanDefinition beanDefinition = (RootBeanDefinition) gatewayProxyBuilder.getBeanDefinition();
+		beanDefinition.setTargetType(serviceInterface);
```

Comment:
You also need to call `setBeanClass` with the factory bean. It is tricky and one of those areas where the core container is not fully consistent. 

Give that a try and let me know how it goes.

Reply:
Didn't help, Stephane.

My code is like this currently:
```
		RootBeanDefinition beanDefinition = (RootBeanDefinition) gatewayProxyBuilder.getBeanDefinition();
		beanDefinition.setBeanClass(GatewayProxyFactoryBean.class);
		beanDefinition.setTargetType(serviceInterface);
```
But an AOT Smoke still fails with the same error: a generated `getControlBusGatewayBeanDefinition()` content is without any changes so far.
I pulled latest `main` and used this command to verify:
```
gradlew clean :integration:build --include-build=..\spring-integration
```
Where `spring-integration` is exactly on this branch with the change we have just discussed.